IntelliTect.TestTools.Data/DatabaseFixture.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Data.Sqlite;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace IntelliTect.TestTools.Data
+{
+    public class DatabaseFixture<TDbContext> : IDisposable where TDbContext : DbContext
+    {
+        private SqliteConnection SqliteConnection { get; }
+
+        private Lazy<DbContextOptions<TDbContext>> Options { get; }
+
+        private IServiceProvider? ServiceProvider { get; set; }
+
+        public DatabaseFixture()
+        {
+            SqliteConnection = new SqliteConnection("DataSource=:memory:");
+            SqliteConnection.Open();
+
+            Options = new Lazy<DbContextOptions<TDbContext>>(() => new DbContextOptionsBuilder<TDbContext>()
+                .UseSqlite(SqliteConnection)
+                .EnableSensitiveDataLogging()
+                .UseLoggerFactory(GetLoggerFactory())
+                .Options);
+        }
+
+        public event EventHandler<ILoggingBuilder>? BeforeLoggingSetup;
+
+        private ILoggerFactory GetLoggerFactory()
+        {
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddLogging(builder =>
+            {
+                builder.AddInMemory();
+
+                BeforeLoggingSetup?.Invoke(this, builder);;
+            });
+
+            ServiceProvider = serviceCollection.BuildServiceProvider();
+
+            return ServiceProvider
+                .GetService<ILoggerFactory>();
+        }
+
+        private TDbContext CreateNewContext()
+        {
+            var constructorInfo = typeof(TDbContext)
+                .GetConstructors()
+                .Where(x =>
+                {
+                    var parameters = x.GetParameters();
+                    return parameters.Length == 1 && parameters[0].ParameterType == typeof(DbContextOptions);
+                })
+                .SingleOrDefault();
+
+            if (constructorInfo is null)
+            {
+                throw new InvalidOperationException(
+                    $"'{typeof(TDbContext)}' must contain a constructor that has a single parameter " +
+                    $"of type '{typeof(DbContextOptions)}'");
+            }
+
+            bool alreadyCreated = Options.IsValueCreated;
+
+            var db = (TDbContext) constructorInfo.Invoke(new object[]{ Options.Value });
+
+            if (!alreadyCreated)
+            {
+                db.Database.EnsureCreated();
+            }
+
+            return db;
+        }
+
+        public async Task PerformDatabaseOperation(Func<TDbContext, Task> operation)
```

Comment:
Perhaps add method documentation specifically calling out the intended use of this method — i.e. to prevent caching, so _not_ around an entire test method.

Reply:
There should definitely be documentation on all public members of this API. Totally dropped the ball on that. Thanks!